addon-test-support/setup-rendering-context.js

```diff
@@ -0,0 +1,135 @@
+import { run, next } from '@ember/runloop';
+import { Promise, resolve } from 'rsvp';
+import Ember from 'ember';
+import jQuery from 'jquery';
+
+export default function(context) {
+  let { owner } = context;
+
+  let dispatcher =
+    owner.lookup('event_dispatcher:main') || Ember.EventDispatcher.create();
+  dispatcher.setup({}, '#ember-testing');
+
+  let OutletView = owner.factoryFor
+    ? owner.factoryFor('view:-outlet')
+    : owner._lookupFactory('view:-outlet');
+  let OutletTemplate = owner.lookup('template:-outlet');
+  let toplevelView = OutletView.create(); // TODO: stash this somewhere?
+  let hasOutletTemplate = !!OutletTemplate;
+  let outletState = {
+    render: {
+      owner,
+      into: undefined,
+      outlet: 'main',
+      name: 'application',
+      controller: context,
+      ViewClass: undefined,
+      template: OutletTemplate,
+    },
+
+    outlets: {},
+  };
+
+  let element, hasRendered;
+  let templateId = 0;
+
+  if (hasOutletTemplate) {
+    run(() => {
+      toplevelView.setOutletState(outletState);
+    });
+  }
+
+  context.render = function render(template) {
+    if (!template) {
+      throw new Error('you must pass a template to `render()`');
+    }
+
+    return resolve().then(function asyncRenderSetup() {
+      let templateFullName = 'template:-undertest-' + ++templateId;
```

Comment:
I understand what it does, but IMHO the following is easier to read and I don't think raw performance is *that* important here

```js
templateId += 1;
let templateFullName = `template:-undertest-${templateId}`;
```